File: sorting/java/DualPivotQuickSort.java
// Java program to implement dual pivot QuickSort (GeeksforGeeks style)
public class DualPivotQuickSort {

    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    private static void dualPivotQuickSort(int[] arr, int low, int high) {
        if (low < high) {
            int[] piv = partition(arr, low, high);
            dualPivotQuickSort(arr, low, piv[0] - 1);
            dualPivotQuickSort(arr, piv[0] + 1, piv[1] - 1);
            dualPivotQuickSort(arr, piv[1] + 1, high);
        }
    }

    private static int[] partition(int[] arr, int low, int high) {
        if (arr[low] > arr[high]) {
            swap(arr, low, high);
        }
        int j = low + 1;
        int g = high - 1;
        int k = low + 1;
        int p = arr[low];
        int q = arr[high];

        while (k <= g) {
            if (arr[k] < p) {
                swap(arr, k, j);
                j++;
            } else if (arr[k] >= q) {
                while (arr[g] > q && k < g) {
                    g--;
                }
                swap(arr, k, g);
                g--;
                if (arr[k] < p) {
                    swap(arr, k, j);
                    j++;
                }
            }
            k++;
        }
        j--;
        g++;
        swap(arr, low, j);
        swap(arr, high, g);
        return new int[] { j, g };
    }

    public static void sort(int[] arr) {
        if (arr == null || arr.length <= 1) {
            return;
        }
        dualPivotQuickSort(arr, 0, arr.length - 1);
    }

    public static void main(String[] args) {
        int[] arr = { 24, 8, 42, 75, 29, 77, 38, 57 };
        sort(arr);
        System.out.print("Sorted array: ");
        for (int v : arr) {
            System.out.print(v + " ");
        }
        System.out.println();
    }
}

File: sorting/python/dual_pivot_quicksort.py
"""
Dual-Pivot QuickSort – versión iterativa (stack explícito).
Evita RecursionError con arreglos de 1 M elementos.
Complejidad: O(n log n) promedio, O(log n) espacio de pila.
"""


def _partition(arr, low, high):
    if arr[low] > arr[high]:
        arr[low], arr[high] = arr[high], arr[low]
    p = arr[low]    # pivote izquierdo
    q = arr[high]   # pivote derecho
    j = k = low + 1
    g = high - 1
    while k <= g:
        if arr[k] < p:
            arr[k], arr[j] = arr[j], arr[k]
            j += 1
        elif arr[k] >= q:
            while arr[g] > q and k < g:
                g -= 1
            arr[k], arr[g] = arr[g], arr[k]
            g -= 1
            if arr[k] < p:
                arr[k], arr[j] = arr[j], arr[k]
                j += 1
        k += 1
    j -= 1
    g += 1
    arr[low], arr[j] = arr[j], arr[low]
    arr[high], arr[g] = arr[g], arr[high]
    return j, g


def dualPivotQuickSort(arr, low=None, high=None):
    """
    Interfaz compatible con el benchmark.
    Si se llama sin low/high, ordena todo el arreglo.
    """
    n = len(arr)
    if n < 2:
        return
    if low is None:
        low = 0
    if high is None:
        high = n - 1

    stack = [(low, high)]
    while stack:
        lo, hi = stack.pop()
        if lo >= hi:
            continue
        lp, rp = _partition(arr, lo, hi)
        # Empujar las tres partes al stack
        stack.append((lo, lp - 1))
        stack.append((lp + 1, rp - 1))
        stack.append((rp + 1, hi))


# Driver
if __name__ == "__main__":
    arr = [24, 8, 42, 75, 29, 77, 38, 57]
    dualPivotQuickSort(arr)
    print("Sorted:", arr)